execute/plugin_functions.go

```diff
@@ -305,31 +305,75 @@ func mergeCommitObservations(
 	return results, nil
 }
 
-// TODO: implement mergeTokenObservations
 func mergeTokenObservations(
-	observations []plugincommon.AttributedObservation[exectypes.Observation],
-	_ map[cciptypes.ChainSelector]int,
-) exectypes.TokenDataObservations {
-	// Finds the observation with 'the most token data'.
-	// This is a dummy algorithm, proper consensus should be implemented.
-
-	maxCnt := -1
-	maxV := exectypes.TokenDataObservations{}
-	for _, ao := range observations {
-		cnt := 0
-		for _, tokenData := range ao.Observation.TokenData {
-			cnt += len(tokenData)
+	aos []plugincommon.AttributedObservation[exectypes.Observation],
+	fChain map[cciptypes.ChainSelector]int,
+) (exectypes.TokenDataObservations, error) {
+	// Single message can transfer multiple tokens, so we need to find consensus on the token level.
+	//nolint:lll
+	validators := make(map[cciptypes.ChainSelector]map[exectypes.MessageTokenID]consensus.MinObservation[exectypes.TokenData])
+	results := make(exectypes.TokenDataObservations)
+
+	for _, ao := range aos {
+		for selector, seqMap := range ao.Observation.TokenData {
+			f, ok := fChain[selector]
+			if !ok {
+				return exectypes.TokenDataObservations{}, fmt.Errorf("no F defined for chain %d", selector)
+			}
+
+			if _, ok1 := results[selector]; !ok1 {
+				results[selector] = make(map[cciptypes.SeqNum]exectypes.MessageTokenData)
+			}
+
+			if _, ok1 := validators[selector]; !ok1 {
+				validators[selector] = make(map[exectypes.MessageTokenID]consensus.MinObservation[exectypes.TokenData])
+			}
+
+			initResultsAndValidators(selector, f, seqMap, results, validators)
 		}
-		if cnt > maxCnt {
-			maxCnt = cnt
-			maxV = ao.Observation.TokenData
+	}
+
+	for selector, seqNrs := range validators {
+		for tokenID, validator := range seqNrs {
+			mtd := results[selector][tokenID.SeqNr]
+			if values := validator.GetValid(); len(values) > 0 {
```

Comment:
In the other consensus functions we've been checking that `len(values) == 1` exactly, idea being if there's more than one value its probably also a consensus error - i.e, there isn't a single value that is "agreed upon". Not sure if this makes sense in the F+1 context which seems to be the current one though.

Assuming we have 3 * F + 1 nodes, you could have up to two different values that have been observed F+1 times, which in my mind is a consensus error. If you do 2*F+1 this isn't possible, but I think for exec we're more lenient than commit.

Reply:
That said, you could just go with this cuz at the end of the day the attestation is verified on-chain, and if an invalid one is chosen the execution will just fail. So its not a big damage.

Reply:
Makes sense, fixing